token/program-2022/src/extension/mod.rs

```diff
@@ -204,6 +204,35 @@ pub struct StateWithExtensionsMut<'data, S: BaseState> {
     tlv_data: &'data mut [u8],
 }
 impl<'data, S: BaseState> StateWithExtensionsMut<'data, S> {
+    /// Resolves a single extension
+    ///
+    /// Warning: performs no validation of the `BaseState`
+    pub fn resolve_single_extension_mut<V: Extension>(
```

Comment:
This `fn` name sucks, but essentially the story of me attempting to modify `process_close_account()` as shown in this PR goes like this:
1. Port `process_close_account()` to use `StateWithExtensionsMut`.  This initially motivated the `pack_base()` change in this PR and it worked great.
2. It worked great until I ran `test_close_account_dups` and it failed on me due to a double borrow, as `StateWithExtensionsMut` holds a mut reference to the account data and this breaks when the close authority is the source account.

(2) caused me to reassess (1) -- porting existing token instructions to use `StateWithExtensionsMut` is risky in general, and instead I really just wanted to resolve a single extension out of a base account that had already been validated.  I didn't want to call `StateWithExtensionsMut::unpack()` on an already unpacked account, and burn more compute units.

wdyt @joncinque?

Reply:
I was thinking that every instruction would get ported to use `StateWithExtensionsMut`, and that we'd be careful about double borrows and drop as needed.  I may be missing something though -- how did this function help your particular situation?  Since `resolve` is doing a mutable borrow on some part of the data anyway, it's still possible to double-borrow at some point.  So this would be something like: `unpack`, do stuff with base + extension, `pack_base`, `drop`, borrow data for something else, `resolve`, do stuff with extension?

Overall, I'd prefer to avoid having this, but if there's a situation that can't get around it, then by all means let's do it.

Man, it's going to be tricky to write tests to cover all of these cases.  Welp, something new to keep me up at night!

Also, let's add a borrow check to `validate_owner` to put at least one worry to rest.

Reply:
> I was thinking that every instruction would get ported to use StateWithExtensionsMut, and that we'd be careful about double borrows and drop as needed

I was hoping we'd only do it on demand actually, and only if strictly needed.  The current token code is battle hardened with some very subtle edge cases. We have decent code coverage but still seems risky.

But for the issue at hand, in the `process_close_account` impl:
1. First the `source_account` is unpacked to figure out if `source_account.owner` should be used as the authority (https://github.com/solana-labs/solana-program-library/blob/0c9f962a7b78be4f99c07a48ea553d3fe76fbb34/token/program-2022/src/processor.rs#L629).  This is the natural place to replace `Account::unpack()` with `StateWithExtensionsMut::unpack()
2. But now `Self::validate_owner` bombs with an immutable borrow of the source_account data.  

This is avoidable in the current `StateWithExtensionsMut`/extensions API by dropping the decoded source account and then unpacking it again, after calling `Self::validate_owner` but that's barf.  Logically at least in this case I'm quite happy to not mess with the bulk of the `process_close_account` code and potentially introduce new bugs, I just want to add one additional simple boring check

Reply:
> I was hoping we'd only do it on demand actually, and only if strictly needed.

Sorry, I should have been clearer.  What I really meant was, "every instruction that is impacted by an extension", and thinking that eventually every instruction would eventually be impacted by an extension.  That may not be the case for awhile though.

> Logically at least in this case I'm quite happy to not mess with the bulk of the process_close_account code and potentially introduce new bugs, I just want to add one additional simple boring check

Sorry, I should have just read the code first.  I late-edited my previous comment instead of adding a new one.  In this particular circumstance, if you use `StateWithExtensionsMut` and move the check up, wouldn't that solve your issue?  As I mentioned I'd prefer to avoid adding this until it's really necessary to avoid a double unpack on the base

Reply:
> What I really meant was, "every instruction that is impacted by an extension", and thinking that eventually every instruction would eventually be impacted by an extension.

I realized half way into my bike ride that every instruction will need to use `StateWithExtensions` after all, as the existing `unpack`s will fail for non-base states! Oh well.

I'll push a new version of this PR with `process_close_account` rewritten with `StateWithExtensionsMut` to demo the issue better later tonight.  I thinking what might help is to break up `StateWithExtensionsMut::unpack` into two things: one that just unpacks the base mint, and one that can be later used to grab a mutable reference to one or more extensions.

Reply:
This thread has gotten a little long, I'll close it and post fresh comments at the new main point of interest